### PR TITLE
Using Webpack to generate our JavaScript loaded CSS

### PR DESCRIPTION
**Pull request by jdlrobson**
This is a precursor for moving our CSS asset building onto node based tooling by focusing first on our JavaScript CSS  assets.
    
See: #1879
See: #1123

### CODE REVIEW

**Comment by cdrini** [comment]
What is "JavaScript based CSS"?

**Comment by jdlrobson** [comment]
Does the retitle help?

**Comment by cdrini** [comment]
Do you think this will have a negative impact on load time? Previously that CSS would start loading once the page's html was loaded and the JS was executed; now the browser has to first make another request for the webpack bundle (which is presumably large) before it starts loading that CSS. Does that sound correct?

**Comment by jdlrobson** [comment]
> Do you think this will have a negative impact on load time? Previously that CSS would start loading once the page's html was loaded and the JS was executed; now the browser has to first make another request for the webpack bundle (which is presumably large) before it starts loading that CSS. Does that sound correct?

It's actually a performance improvement. Currently loadStyle creates and inserts a style tag which is downloaded in parallel to the JS tag. Now the styles are bundled in the js file so we make one less request. The styles are useless without the JS so before if JS loaded before CSS or even worse only one of them loaded we were shipping a broken experience. We are already making a request for the webpack bundle...it just now includes the css too (yay!)

**Comment by cdrini** [comment]
Oh, ok; I didn't realize the CSS would now be directly included in the bundle! That makes more sense :P So we're reducing the number of requests, but increasing the size of the bundle; could you by any chance provide some performance metrics to show this is net faster or the same? I think the JS is probably more important (since the CSS will (presumably) only be needed once that's loaded/executing), and I want to make sure that it isn't delayed by the extra bundle size.

**Comment by cdrini** [comment]
Huh, this is strange; the css looks like it starts loading _after_ the js:

image
_With caching; no throttling_

image
_Cache disabled; throttling: Regular 3G_

Is this because of the `window.q` stuff?

**Comment by jdlrobson** [comment]
js-all.css should not be loaded at all. I'm not sure why you are seeing this in the network tab. Did you run make js and clear memcache?

> could you by any chance provide some performance metrics to show this is net faster or the same? 

Previously: 
- page-<page>.css was loaded at the top of the page, blocking first paint
- all.js was loaded  at the end of the page, blocks fully load time but delays "first interactive"
- all.js pulled in js-all.css using the loadStyle function and an additional HTTP request

Now:
- page-<page>.css was loaded at the top of the page, blocking first paint
- all.js loads at the end of the page, blocks fully load time but delays "first interactive" and contains  the contents of js-all.css

Since all.js and js-all.css should ideally be loading at the same time I am not sure what performance metrics would show.  In the previous setup there is a moment of time between all.js and js-all.css loading where the page is interactive (e.g. I can click "change cover" under a book) but the styles are not (the modal window would not be styled). This is obviously problematic and what I'm looking to remove. The main metric we should be caring about is first paint and that isn't impacted by this change.

**Comment by cdrini** [comment]
Oh, sorry, I should have mentioned those are from prod; I just wanted to get a benchmark :P

I think first paint will be unaffected, which is fine; I'm concerned this might affect first interactive, which we should try to avoid. The current (on prod) solution is a little flawed; the line that loads the CSS uses the `loadStyle` function, so it must happen _after_ the bundle has been loaded (which is why it's delayed in the network log). Ideally this would start loading with all the other resources. This would reduce the total load time to `~7.5s` (from the current `10.5`). Note there is a pretty jarring styling flash (mainly the slider buttons), so reducing the total time is also important.

---

> The main metric we should be caring about is first paint and that isn't impacted by this change.

I don't think we should ignore other metrics, though. Here's the performance metrics which I think are important:
1. first paint, but this shouldn't be affected, so it doesn't really matter
2. first interactive (when is all the JS loaded)
3. last ui change (when is the last visual change in the UI)
4. total time to ready (when are all the resources finished loading)

The last 3 are the ones I'm curious about.

---

> there is a moment of time between all.js and js-all.css loading where the page is interactive (e.g. I can click "change cover" under a book) but the styles are not (the modal window would not be styled)

I think this ok, because it requires user input. This means the CSS _can_ take a little longer to load, because the user is unlikely to click it in the window of time before the CSS is done loading. But if we force the JS/CSS to be loaded/bundled together, then they both have to be ready, so first-interactive, last-ui-change, and total-time will all be affected. Also because the browser can no longer load the JS/CSS in parallel; it now has to load them sequentially.

---

Basically, could you provide network diagrams (like above) for 3 scenarios:
1. The current setup
2. The setup in this PR
3. Inlining the `loadStyle` function into the HTML file so the CSS can start loading at the same time as the JS; that should have a big impact on metrics 3 and 4

My hypothesis is that this PR is probably more performant than the current setup, but less performant than option #3.

**Comment by jdlrobson** [comment]
But 3 would create cacheability problems and increase the size of the HTML file which in turn would delay the loading of JS and CSS. That's the same problem.

To be clear this proposed patchset will delay the time the JS loads/first interactive (by a small percentage) but that is misleading as right now the first interactive is actually flawed as it has no associated CSS loaded so actually first interactive is when all-js loads which this improves on.

This patch will thus improve time to load and time to first interactive and I can prove that to you only after it has been merged (we can also run webpagetest on dev if we desire but the environment is too unstable performance wise to be relied on as an indication of what it will be in production)

Happy to talk through this in slack as i fear we are getting our wires crossed somewhere.

**Comment by cdrini** [comment]
For 3, I mean inline the _function_ not the CSS. The function is like 8 lines, so it won't have a meaningful impact on HTML load time. I'm thinking you could take snapshots of the network panel running from your local environment. It's not perfect/identical to prod, but it'll give a good enough approximation without having to send code to prod.

I'll message you on slack :)

**Comment by cdrini** [comment]
Summary of chat: According to https://www.webpagetest.org/result/190329_P3_a3d0a21f8262a3f95e5e26bdd3cf5284/ , that CSS takes significantly shorter to load (unlike the FF logs above, for some reason), so parallelizing the loading of the two would probably have little effect, and bundling it with the JS shouldn't have a very large effect on its size. If I get the time, I'd still like to make some network tab comparisons, but over all this makes sense to me :)
